experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts

```diff
@@ -0,0 +1,193 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/core-utils";
+
+import {
+	EmptyKey,
+	rootFieldKey,
+	type FieldKey,
+	type TreeTypeSet,
+	type Value,
+	CursorLocationType,
+} from "../../core";
+// eslint-disable-next-line import/no-internal-modules
+import { leaf } from "../../domains/leafDomain";
+import {
+	type ContextuallyTypedNodeData,
+	type TreeDataContext,
+	getPossibleTypes,
+	isFluidHandle,
+	isPrimitiveValue,
+} from "../contextuallyTyped";
+import {
+	type CursorAdapter,
+	type CursorWithNode,
+	stackTreeNodeCursor,
+	stackTreeFieldCursor,
+} from "../treeCursorUtils";
+import { TreeNodeSchema } from "../typed-schema";
+import { type ProxyNode } from "./types";
+/**
+ * Creates a {@link CursorAdapter} capable of mapping proxy-domain trees to the editable tree domain.
+ *
+ * @param context - The {@link TreeEntity.context} of the parent node under which the tree is being inserted.
+ * @param typeSet - The types allowed under the parent node under which the tree is being inserted.
+ *
+ * @privateRemarks
+ * This is exported solely for testing purposes.
+ * Consumers should use {@link cursorFromProxyTree} instead.
+ */
+export function createProxyTreeAdapter<TNode extends ProxyNode<TreeNodeSchema, "javaScript">>(
+	context: TreeDataContext,
+	typeSet: TreeTypeSet,
+): CursorAdapter<TNode> {
+	return {
+		value: (node: TNode) => {
+			return isPrimitiveValue(node) || node === null || isFluidHandle(node)
+				? (node as Value)
+				: undefined;
+		},
+		type: (node: TNode) => {
+			if (node === undefined) {
+				throw new Error(
+					"Cursor adapter was passed an undefined node. This is unsupported.",
+				);
+			}
+			if (node === null) {
+				return leaf.null.name;
+			}
+
+			switch (typeof node) {
+				case "number":
+					return leaf.number.name;
+				case "string":
+					return leaf.string.name;
+				case "boolean":
+					return leaf.boolean.name;
+				default: {
+					if (isFluidHandle(node)) {
+						return leaf.handle.name;
+					}
+
+					const possibleTypes = getPossibleTypes(
```

Comment:
At this point you have a ProxyNode right? Don't you just want Tree.schema(node).name here?

Reply:
`node` is a `ProxyNode<"javaScript">`, not a `ProxyNode<"sharedTree">`. Only the latter can be used in such a way. The stuff being processed here is essentially POJO.

Reply:
I'll give that a shot. I tried using that at one point and it didn't work, but that was early on.